index.go

```diff
@@ -242,15 +267,56 @@ func LoadIndexXML(r io.Reader) (*Index, error) {
 	return &index, nil
 }
 
-func (a *App) CurApk() *Apk {
+func (a *App) ApksByVName(vname string) []Apk {
+	var apks []Apk
+	for i := range a.Apks {
+		if vname == a.Apks[i].VName {
+			apks = append(apks, a.Apks[i])
+		}
+	}
+	return apks
+}
+
+func (a *App) SuggestedVName() string {
 	for i := range a.Apks {
 		apk := a.Apks[i]
 		if a.CVCode >= apk.VCode {
+			return apk.VName
+		}
+	}
+	return ""
+}
+
+func (a *App) SuggestedApks() []Apk {
+	// No APKs => nothing to suggest
+	if len(a.Apks) == 0 {
+		return nil
+	}
+
+	// First, try to follow CV
+	apks := a.ApksByVName(a.SuggestedVName())
+	if len(apks) > 0 {
+		return apks
+	}
+
+	// When CV is missing current version code or it's invalid (no APKs
+	// match it), use heuristic: find all APKs having the same version
+	// string as the APK with the greatest version code
+	return a.ApksByVName(a.Apks[0].VName)
+}
+
+func (a *App) SuggestedApk(device *adb.Device) *Apk {
+	for i := range a.Apks {
+		apk := a.Apks[i]
+		if a.CVCode >= apk.VCode && apk.IsCompatible(device) {
 			return &apk
 		}
 	}
-	if len(a.Apks) > 0 {
-		return &a.Apks[0]
+	for i := range a.Apks {
+		apk := a.Apks[i]
+		if apk.IsCompatible(device) {
+			return &apk
```

Comment:
Note that you're returning the address of a local variable here, not the address of the struct in the slice. You probably want:

```go
for i := range a.Apks {
	apk := &a.Apks[i]
	if apk.IsCompatible(device) {
		return apk
	}
}
```